lnrpc/routerrpc/router_server.go

```diff
@@ -656,3 +669,192 @@ func (s *Server) SubscribeHtlcEvents(req *SubscribeHtlcEventsRequest,
 		}
 	}
 }
+
+// htlcInterceptInfo is used in the HtlcInterceptor bidirectional stream and
+// encapsulates the request information sent from the HtlcForwardHandler to the
+// RPCServer.
+type htlcInterceptInfo struct {
+	inKey        channeldb.CircuitKey
+	htlc         lnwire.UpdateAddHTLC
+	responseChan chan bool
+}
+
+func (s *Server) HtlcInterceptor(stream Router_HtlcInterceptorServer) error {
+
+	// Create two channels to handle requests and responses respectively.
+	newRequests := make(chan *htlcInterceptInfo)
+	responses := make(chan ForwardHtlcInterceptResponse)
+
+	// Define a quit channel that will be used to signal to the HtlcForwardHandler
+	// whether the stream still exists.
+	quit := make(chan struct{})
+	defer close(quit)
+
+	// forwardHandler coordinates between the htlcinterceptor.Middleware and the
+	// RPC responses. It intercepts the Middleware requests, wait for the RPC response
+	// and send a synchronous result to the Middleware.
+	forwardHandler := func(inKey channeldb.CircuitKey,
+		htlc lnwire.UpdateAddHTLC) bool {
+
+		respChan := make(chan bool, 1)
+		newRequest := &htlcInterceptInfo{
+			inKey:        inKey,
+			htlc:         htlc,
+			responseChan: respChan,
+		}
+
+		// timeout is the time after which ForwardHtlcInterceptRequest expire.
+		timeout := time.After(defaultForwardInterceptionTimeout)
+
+		// Send the request to the newRequests channel.
+		select {
+		case newRequests <- newRequest:
+		case <-timeout:
+			log.Errorf("Forward handler returned false - reached timeout of %d",
+				defaultForwardInterceptionTimeout)
+			return false
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+
+		// Receive the response and return it. If no response has been received
+		// in defaultForwardInterceptionTimeout, then return false.
+		select {
+		case resp := <-respChan:
+			return resp
+		case <-timeout:
+			log.Errorf("Forward handler returned false - reached timeout of %d",
+				defaultForwardInterceptionTimeout)
+			return false
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+	}
+
+	// Add our forwardHandler to the Middleware.
+	interceptMiddleware := s.cfg.RouterBackend.HtlcInterceptMiddleware
+	handlerID := interceptMiddleware.AddForwardHtlcHandler(forwardHandler)
+	defer interceptMiddleware.RemoveForwardHtlcHandler(handlerID)
```

Comment:
> I think we still need to cancel back any outstanding HTLC resolution intents.

I understand the need to do some action for outstanding HTLC , I just wondering if the default behavior shouldn't be `Resume` rather than `Cancel`.
This is mainly because I don't see a different between timeout, disconnection or no interceptor at all.